### ISSUE TITLE
Apply differential testing based on a reference implementation

### ISSUE THREAD

**Issue by danwt** [cosmos/interchain-security#60]
# Idea

## Use a reference implementation to generate correct states from actions

It might be feasible to take a property based testing approach. We could implement the relevant system in very simple, single threaded, in-memory, imperative code without any regards to performance (this is the 'model') and execute actions against it in order to get correct states for each action. The actions could be generated dumb-randomly or smart-randomly (like quickcheck ect) or something inbetween. Then we combine the actions and states, forming traces, and run the traces MBT style.

Note: there might be some existing tools we could use here, maybe even leveraging an existing PBT framework.

![pbt](https://i.imgur.com/BAHI4Zv.png)

## Wider context

I think this could be a good thing to do first as by implementing a reference implementation(s) I would learn a lot and this would help me write the TLA+ models later, which will be used for traditional verification (@mpoke) and also regular MBT. This work would also complement efforts to create a better test driver (see e.g. https://github.com/cosmos/interchain-security/issues/58#issuecomment-1106715041) as a driver will be needed to execute traces. That interface of the driver could hook into e.g. a fork of the ibc-go testing framework.

(I discussed this idea with @jtremback earlier today).

![wider context](https://i.imgur.com/JFb4sbs.png)

## Questions

The key question is whether it is cost effective to write a reference implementation. It **must** be protocol correct, of course! My feeling is that it will be very doable.

**Comment by jtremback** [comment]
I think that in reference to both this issue and https://github.com/cosmos/interchain-security/issues/61, the important part will be to implement a driver that takes json traces and applies them to either a real network (like our integration tests) or the IBC-Go simulated chains. 

This will be the bulk of the work, and from here it will be very easy to test in a variety of different ways, from hand written traces, to trivially randomized traces that just try to break invariants ("fuzzing"?), to property based or model based testing.

I would prioritize writing the driver, unless you think that the driver would be much easier with the knowledge gained from writing a reference implementation.

**Comment by jtremback** [comment]
Actually, since we already have a driver for the integration tests, maybe an attempt to drive that with a reference implementation would be a good way to highlight any issues in the driver, and also help us learn how useful tests on a real network are vs tests on the IBC-Go test framework.

**Comment by mpoke** [comment]
I like the idea of a reference implementation, but I don't know how difficult would be to make one that contains enough details to be relevant. For example, we could implement the entire IBC communication as a queue, but a straightforward implementation would not cover relaying delays.

**Comment by danwt** [comment]
@sainoe this is the best issue for describing the methodology we discussed in our meeting

**Comment by mpoke** [comment]
The action space for CCV (i.e., external API):

- on provider 
  - delegate(D, V, A), where D is a delegator, V is a validator and A is an amount
  - undelegate(D, V, A)
  - redelegate(D, Vsrc, Vdst, A)
  - slash(V, infractionHeight, slashFactor)
  - jumpToBlock(h), where h is a height larger than the current height
- on consumer
  - slash(V, infractionHeight, slashFactor)
  - jumpToBlock(h)

As discussed with @danwt, for now we assume that the channel initialization is established.

**Comment by danwt** [comment]
> I like the idea of a reference implementation, but I don't know how difficult would be to make one that contains enough details to be relevant. For example, we could implement the entire IBC communication as a queue, but a straightforward implementation would not cover relaying delays.

From looking at what ibc-go/testing offers, and what hermes offer, in terms of testing it should be easy to model the network as a queue with `pushOne(x)` and `popAll()`, but it might be harder to implement `popOne`. This is because `sendPacket` on the sender will be like `pushOne` and `UpdateClient` on the recipient chain will do `popAll()` (same for ` hermes tx raw packet-recv`).

**Comment by konnov** [comment]
Hey @danwt, it just occurred to me that we might have missed the following option in the discussions. How about:

 1. Using TLC in the simulation mode to produce random traces,
 2. converting them to ITF and
 3. executing them with Atomkraft or your custom driver?

I don't see any drawbacks in comparison to PBT here. You would get non-determinism and expressiveness of TLA+ for free. Obviously, you would have the spec in TLA+, not a reference model in Python.

**Comment by danwt** [comment]
Hey @konnov, I did as you describe for the staking module in Q1 and it did work just fine. However, I have not yet made a judgement on what is the better choice between a) using TLA+ and b) using a programming language (e.g Python). Here are some pros and cons so far, which I try to quantify. I compare to Python but most any programming language would do.

### Advantages of TLA+ over, say, Python

- In principle you can model check the spec in order to check that you wrote it correctly. In Python you are limited to software testing. In _practice_ though with the technique described, the model is too big to model check (that's why you run simulation!)
- TLA+ is expressive, you can easily describe a set of initial states, or a set of transitions by using e.g. `x \in [S -> T]` where `S` and `T` can themselves be expressive expressions. In Python you have to right this by hand. In _practice_ though\
    a) It can be hard to choose initial states far from the 'absolute zero' state because you must know that any of the state are actually reachable from the 'absolute zero' state. Also, given an unlimited number of actions, it is actually sufficient to use the 'absolute zero' state as the initial state. Doing so is only undesirable when you are model checking and thus limited in the number of actions.
    b) The on-chain algorithms can be represented by a determintic state machine and all non determinism is confined to the choice of input word. The input word is (so far in my experience) made of simple API calls which can easily be [randomized](https://docs.python.org/3/library/random.html) in Python (`Delegate(val=randrange(k),amt=randrange(10))`)
    c) In my opinion it is usually faster and easier to write the logic snippets encountered in the business logic of the algorithms being modelled in Python than TLA+. I also think the Python is probably easier to modify than equivalent TLA+. I can provide many examples if needed but as an obvious one compare `sum(x.foo for x in S)` to equivalent TLA+. My hunch is that most people who are good at writing TLA+ would agree with this point.


### Advantages of, say, Python over TLA+

In evaluating the advantages of TLA+ I quantified with some advantages of Python, so I try not to repeat here.

- Software Engineering\
    It is easier to debug, write unit tests for, and use functionality written by others (packages) in Python than in TLA+. I would say it's easier write in general too (point _c_ above). Moreover, the Python interactive development experience offered by VSCode or other tools is strictly better than the TLA+ equivalent (I hope this can change!).
- Community\
    Anyone can read, understand, review and modify Python code. The same cannot be said for TLA+. It is easier to hire someone to take over ownership of Python code than TLA+ code when the original author is no longer available.
- Probability distributions\
    In TLA+ it is easy to restrict the set of possible actions in each step but it would not be easy to change the probability distribution. In Python you can concisely write logic for both.
- Writing additional arbitrary code that extends possiblities is easier in Python than TLA+ due to the _Software Engineering_ aspect described above. As an example consider I have an action than progresses the blockchain by several blocks. In Python I can easily save each block state in the trace in addition to actual _model state_ which will have jumped several blocks. In TLA+ this would be difficult.

At this point in time I'm more inclined to favour Python, based on the analysis above. I would be very interested for someone to demonstrate enough effectiveness using TLA+ to change my mind.

**Comment by danwt** [comment]
Closed as PR

- https://github.com/cosmos/interchain-security/pull/126

tackles this partially but now it is blocked by the TDD dev cycle. That is: the tests in the PR fail due to a the SUT not being fully implemented. I'm not sure if it makes sense to merge the PR currently, but further work is fine grained and deserves its own issue. Furthe more, this issue is not well written with closing criteria ect.
